### PR TITLE
Halt compaction if job is lost

### PR DESCRIPTION
**Pull request by joe-elliott**
**What this PR does**:
In a [previous PR](https://github.com/grafana/tempo/pull/4296) we added a log message when compactors finished a job and recognized it no longer owned the job. We have since seen this message pop up sporadically in production suggesting that unstable compactors will rarely increase the amount of data in object storage.

This PR extends on that work by adding code to abandon compaction in the event that ownership of the job changes. This is not foolproof but should make this an extremely rare occurrence.

~I also considered an approach where we created a child context in `doCompaction` and cancel it with cause in the event that job ownership changes. We could use the cause to confirm that a "context cancelled" that bubbles up from `.compact()` was in fact due to ownership change. The reason I don't like it is b/c we would have to create a side goroutine that polls `.Owns()` watching for ownership change which felt messier. I am definitely game to go this route if others prefer it.~

I have since pivoted to the approach described in the strikethrough for the reason mentioned below. The PR now contains the following changes:

- Cleaned up the compaction loop to use a simple `time.Sleep` to accomplish the same intent.
- Made the start and end compaction cycle log message "Info" level like other status updates for consistency
- Removed <-ctx.Done() and just checked ctx.Err() for simplicity in doCompaction.
- Renamed `doCompaction` and `compact` for clarity and added a `compactWhileOwns` method that does the job of checking job ownership during compaction

I have also left the first attempt as the first commit in this branch and then reverted it if a reviewer would like to see both options.

**Checklist**
- [x] Tests updated
- [ ] Documentation added
- [x] `CHANGELOG.md` updated - the order of entries should be `[CHANGE]`, `[FEATURE]`, `[ENHANCEMENT]`, `[BUGFIX]`

### CODE REVIEW

File: tempodb/compactor.go [hunk 1]
```diff
@@ -142,51 +136,99 @@ func (rw *readerWriter) doCompaction(ctx context.Context) {
...
-			if start.Add(rw.compactorCfg.MaxTimePerTenant).Before(time.Now()) {
-				measureOutstandingBlocks(tenantID, blockSelector, rw.compactorSharder.Owns)
+	done := make(chan struct{})
+	defer close(done)
 
-				level.Info(rw.logger).Log("msg", "compacted blocks for a maintenance cycle, bailing out", "tenantID", tenantID)
+	// every second test if we still own the job. if we don't then cancel the context with a cause
+	// that we can then test for
+	go func() {
+		ticker := time.NewTicker(1 * time.Second)
```
Comment (zalegrala) [inline tempodb/compactor.go:192]: Defer the ticker stop?

File: tempodb/compactor.go [hunk 2]
```diff
@@ -142,51 +136,99 @@ func (rw *readerWriter) doCompaction(ctx context.Context) {
...
+
+		for {
+			if !owns() {
+				cancel(errCompactionJobNoLongerOwned)
```
Comment (zalegrala) [inline tempodb/compactor.go:196]: Return after the cancel here?

File: tempodb/compactor.go [hunk 3]
```diff
@@ -380,3 +422,21 @@ func (i instrumentedObjectCombiner) Combine(dataEncoding string, objs ...[]byte)
 	}
 	return b, wasCombined, err
 }
+
+// doForAtLeast executes the function f. It blocks for at least the passed duration but can go longer. if context is cancelled after
+// the function is done we will bail immediately. in the current use case this means that the process is shutting down
+// we don't force f() to cancel, we assume it also responds to the cancelled context
+func doForAtLeast(ctx context.Context, dur time.Duration, f func()) {
```
Comment (zalegrala) [inline tempodb/compactor.go:429]: Hmm, I'm trying to understand this.  It looks like this still only executes `f` once no matter the context status.  What am I overlooking?

File: tempodb/compactor.go [hunk 4]
```diff
@@ -380,3 +422,21 @@ func (i instrumentedObjectCombiner) Combine(dataEncoding string, objs ...[]byte)
...
+	startTime := time.Now()
+	f()
+	elapsed := time.Since(startTime)
+
+	if elapsed < dur {
+		ticker := time.NewTicker(dur - elapsed)
```
Comment (zalegrala) [inline tempodb/compactor.go:435]: Defer the ticker stop.

File: tempodb/compactor.go (hunk 3 above)
Comment (zalegrala) [reply tempodb/compactor.go:429]: Oh we're holding the blocked call to f for at least duration.

File: tempodb/compactor.go [hunk 5]
```diff
@@ -380,3 +422,21 @@ func (i instrumentedObjectCombiner) Combine(dataEncoding string, objs ...[]byte)
...
+
+		select {
+		case <-ticker.C:
+		case <-ctx.Done():
```
Comment (zalegrala) [inline tempodb/compactor.go:439]: Return here?

File: tempodb/compactor.go (hunk 1 above)
Comment (joe-elliott) [reply tempodb/compactor.go:192]: good catch. 🙏

File: tempodb/compactor.go (hunk 2 above)
Comment (joe-elliott) [reply tempodb/compactor.go:196]: i thought i'd just let it drop down to the select which would bail out, but i like your suggestion. it's more clear.

File: tempodb/compactor.go (hunk 5 above)
Comment (joe-elliott) [reply tempodb/compactor.go:439]: both cases will just fall through to the end of the function. this basically just blocks until one of the two things occurs.

File: tempodb/compactor.go (hunk 3 above)
Comment (joe-elliott) [reply tempodb/compactor.go:431]: it was meant to replicate the current logic in a more straight forward way. i'm not 100% sure, but the current logic:
```
	for {
		select {
		case <-ctx.Done():
			return
		default:
		}

		select {
		case <-ticker.C:
			rw.doCompaction(ctx)
		case <-ctx.Done():
			return
		}
	}
```

Basically just attempts to run the `doCompaction` function once every `CompactionCycle`. Also, if compaction takes a long time then the ticker would have elapsed potentially hundreds of times over. I wonder if this is related to our go 1.23 timer panic?

So I tried to write a simpler method that:
- ran the compaction function
- if it took more than the CompactionCycle then immediately return so we can do another
- if it took less time than the CompactionCycle then wait for the difference while also watching for a context cancel.

File: tempodb/compactor.go (hunk 5 above)
Comment (zalegrala) [reply tempodb/compactor.go:442]: Yeah, its interesting but no issues.

File: tempodb/compactor.go [hunk 6]
```diff
@@ -142,51 +136,101 @@ func (rw *readerWriter) doCompaction(ctx context.Context) {
...
 	start := time.Now()
 
-	level.Debug(rw.logger).Log("msg", "starting compaction cycle", "tenantID", tenantID, "offset", rw.compactorTenantOffset)
+	level.Info(rw.logger).Log("msg", "starting compaction cycle", "tenantID", tenantID, "offset", rw.compactorTenantOffset)
 	for {
-		select {
-		case <-ctx.Done():
+		// this context is controlled by the service manager. it being cancelled means that the process is shutting down
+		if ctx.Err() != nil {
+			level.Info(rw.logger).Log("msg", "caught context cancelled at the top of the compaction loop. bailing.", "err", ctx.Err(), "cause", context.Cause(ctx))
```
Comment (zalegrala) [inline tempodb/compactor.go:143]: Should we return here?

File: tempodb/compactor.go (hunk 3 above)
Comment (zalegrala) [reply tempodb/compactor.go:431]: This is clever and I like the approach.  It took a minute to see the reason, but holding a wait period for the cancellation makes sense.

File: tempodb/compactor.go [hunk 7]
```diff
@@ -142,51 +136,101 @@ func (rw *readerWriter) doCompaction(ctx context.Context) {
...
+func (rw *readerWriter) compactWhileOwns(ctx context.Context, blockMetas []*backend.BlockMeta, tenantID string, owns func() bool) error {
+	ownsCtx, cancel := context.WithCancelCause(ctx)
+
+	done := make(chan struct{})
+	defer close(done)
+
+	// every second test if we still own the job. if we don't then cancel the context with a cause
+	// that we can then test for
+	go func() {
+		ticker := time.NewTicker(1 * time.Second)
```
Comment (zalegrala) [inline tempodb/compactor.go:192]: I see now it appears this 1 second is in conflict with the `compactionCycle`.  If we have a `compactionCycle` less than 1 second, it means we could continue compaction for one extra than desired.  Is this a concern?

**Review by zalegrala — COMMENTED** [review]
This is neat.  Just a couple comments.

File: tempodb/compactor.go (hunk 6 above)
Comment (joe-elliott) [reply tempodb/compactor.go:143]: wups, yes

File: tempodb/compactor.go (hunk 7 above)
Comment (joe-elliott) [reply tempodb/compactor.go:192]: this doesn't block the return of the function so it won't stop the compaction. if a compaction takes < 1s then this method will still return successfully and some short amount of time later this goroutine will exit.

**Review by zalegrala — APPROVED** [review]

**Comment by Wessie** [comment]
Hello, I just hit a code path introduced in this patch that made tempo spin a CPU core.

This is because the for loop on [this line](https://github.com/grafana/tempo/blob/main/tempodb/compactor.go#L86) has no exit condition. This means when tempo is exiting (but slowly for some reason), the `ctx` is canceled and this makes `f()` instantly return with this kind of log message:

`caller=compactor.go:143 msg="caught context cancelled at the top of the compaction loop. bailing." err="context canceled" cause="context canceled"` 

and then [this line](https://github.com/grafana/tempo/blob/main/tempodb/compactor.go#L443) also returns instantly inside of `doForAtLeast` because `ctx` is canceled. Causing the problem.

**Comment by joe-elliott** [comment]
@Wessie Great find. Thank you 🙏. Would you like to submit a PR to fix?

If not I will patch it up today, no worries.

**Comment by Wessie** [comment]
Feel free to patch.

**Pull request merged**